### PR TITLE
Proposal: A way to release Openmrs SPAs

### PR DESCRIPTION
**Pull request by nkimaina**

### CODE REVIEW

File: text/0021-versioning-releases.md
Comment (ibacher) [inline text/0021-versioning-releases.md]: "fork" rather than "folk". I'm not clear on what the "for the purposes of releasing ONLY" clause is adding. "to diverge, Openmrs..." should probably read "to diverge from Openmrs, not benefiting...".

File: text/0021-versioning-releases.md
Comment (ibacher) [inline text/0021-versioning-releases.md]: "open-source projects" -> "SPA applications"

File: text/0021-versioning-releases.md
Comment (ibacher) [inline text/0021-versioning-releases.md]: "Have a defined long-term support window for releases"

File: text/0021-versioning-releases.md
Comment (ibacher) [inline text/0021-versioning-releases.md]: We should probably have a stronger reason for landing on SemVer than just "it is the most widely accepted versioning scheme". Among other things, this decision is both central to the point of this RFC and a break with how the OpenMRS community currently manages versions, though arguably OpenMRS as a whole should move towards adopting SemVer.

I would favour something like: "as SemVer allows implementors to more easily assess the work involved in upgrading an application".

File: text/0021-versioning-releases.md
Comment (ibacher) [inline text/0021-versioning-releases.md]: "digital ocean" -> "Digital Ocean". Do we want an RFC that refers to what the OpenMRS SPA test environment is "currently" doing or even to this environment at all? I think it's sufficient to say "Releases should be tagged and the artifacts distributed via a CDN maintained by OpenMRS, so that implementors have stable resources to choose from".

File: text/0021-versioning-releases.md
Comment (ibacher) [inline text/0021-versioning-releases.md]: "SPA applications" is redundant. "can have control on" -> "has control over". "when" -> "when and whether". "the latest changes" -> "more recent versions".

File: text/0021-versioning-releases.md
Comment (ibacher) [inline text/0021-versioning-releases.md:23]: At which level should these maintenance branches be maintained? Should a SPA maintain a branch for every minor version or only for major versions?

File: text/0021-versioning-releases.md
Comment (ibacher) [inline text/0021-versioning-releases.md]: "In addition to the reason 1) above" - remove this text

File: text/0021-versioning-releases.md
Comment (ibacher) [reply text/0021-versioning-releases.md:23]: Should this RFC make a decision on how bugfixes are addressed vis-a-vis more recent versions? I.e., should we favour bugfixes against 2.x which are backported to 1.x (this is something like the current OpenMRS model) or bugfixes against 1.x which are "forward-ported" to 2.x (this involves more work for the maintainers, but makes it easier for implementors to provide bugfixes).

**Review by ibacher — COMMENTED** [review]
1. This document uses different capitalisations of both OpenMRS and SPA. These should be unified. I'd favour "OpenMRS" and "SPA".

1. "OpenMRS" is used as a term to variously refer to (a) the OpenMRS community and (b) whatever we are calling the version of the OpenMRS platform + the Single SPA module (is there an accepted name for this?). We should clearly distinguish these two things.

File: text/0021-versioning-releases.md
Comment (mseaton) [reply text/0021-versioning-releases.md]: @ibacher OpenMRS has standardized on SemVer I believe, it just might not be fully enforced or applied retroactively to everything

Comment (ibacher) [reply text/0021-versioning-releases.md]: @mseaton Fair. I still think the RFC can provide a stronger rationale for this decision.

File: text/0021-versioning-releases.md
Comment (brandones) [reply text/0021-versioning-releases.md:23]: If we are going to do release branching, it should only be for major versions.

**Comment by brandones** [comment]
This is about versioning ESMs? They are already versioned according to semvar.

We do not have release branches, and I do not think that we should take on the responsibility of backporting bug fixes, which will reduce our velocity, until there are multiple organizations using MFE in production, and at least one of them decides that they want to stay on an old version of some ESM. That is the point at which we should be considering this.

File: text/0021-versioning-releases.md
Comment (nkimaina) [reply text/0021-versioning-releases.md:23]: I think we should have release branches for both, as the whole point is to allow implementors to decide when to upgrade to see new features, as well as allow them to benefit from bug-fixes.

Comment (nkimaina) [reply text/0021-versioning-releases.md:23]: The alternative will be to do feature flags, which I think will get messy for an open-source project.

Comment (brandones) [reply text/0021-versioning-releases.md:23]: The kinds of features that people would generally want to disable should be toggled in the config.

For everything else, organizations should ask themselves -- "are we really that special?" And if so, they should introduce a new config parameter.

Backporting bug fixes to every minor version:
- is not justifiable based on current adoption patterns
- will be a huge pain for developers
- will discourage us from releasing, knowing that every release is a new maintenance nightmare we will have to live with until it is EOL'd
- will reduce forward velocity on MFE before it has even had the opportunity to prove itself, at a time when organizations still consider dedicating devs to MFE a risky proposition

For major versions, there is much more of an argument to be made for release branching, since the definition of a major version is one that introduces breaking changes -- i.e., the upgrade requires some effort, and organizations often take a while to make that happen.

Comment (nkimaina) [reply text/0021-versioning-releases.md:23]: If it was library that we are building, then what you are saying makes perfect sense. These front-end features are client-facing, users interact with them, there's training that goes with it, and decisions about workflows that are made by various stakeholders before a feature is introduced. We need to allow an organisation to benefit from bug-fixes and patches, before they are ready to let their users use a new feature or a modification of an existing feature. Effort to back-port bugs to certain release-branches could be left to such an organisation. Configs to support 'timing of release of a certain feature to a certain group of users' will require coordination each release cycle, and effort to maintain anyway, and can get ugly for an open-source project. 
All this said, I think it would be safer to allow creation of these minor release branches on demand, and leaving the responsibility of maintaining it to an organisation that needs it badly. Restricting it will only promote organisations to create forks when they get cornered with a bug that affects them.

Comment (nkimaina) [reply text/0021-versioning-releases.md:23]: We still need the configs that are there by design, to allow an application to be configured to an organisation's  needs, but requiring that every feature to have these configs for the purposes of release adds complexities to the entire application.

File: text/0021-versioning-releases.md
Comment (brandones) [inline text/0021-versioning-releases.md]: Since most MFEs don't have an API, I'm going to suggest

```suggestion
MAJOR Version: version where there are incompatible API or configuration API changes, or major UI changes that are enabled by default
```

File: text/0021-versioning-releases.md
Comment (brandones) [reply text/0021-versioning-releases.md:23]: Really good points, @nkimaina . I'm in agreement with letting organizations create and maintain release branches according to their needs.

I'm also going to propose some changes to our definition of major and minor versions, which might help accommodate the needs you are designing for.

File: text/0021-versioning-releases.md
Comment (brandones) [inline text/0021-versioning-releases.md:21]: I'd delete this line -- SPA Applications already are versioned according to SemVar.

File: text/0021-versioning-releases.md
Comment (brandones) [inline text/0021-versioning-releases.md]:
```suggestion
The OpenMRS Microfrontend Squad will not maintain release branches. Implementers 
or organizations who wish to maintain release branches may do so. They 
should be granted write access to the GitHub repo hosting the ESM of interest.
```

File: text/0021-versioning-releases.md
Comment (brandones) [inline text/0021-versioning-releases.md]: I think we should feel free to version and release as often as our progress warrants. If we're getting a lot of requests to "please release X," that will mean we are not releasing often enough. I would just get rid of this line, since what I'm suggesting is what we already do.

File: text/0021-versioning-releases.md
Comment (brandones) [inline text/0021-versioning-releases.md]: We do use semvar.

File: text/0021-versioning-releases.md
Comment (brandones) [inline text/0021-versioning-releases.md]: Can we be specific about when release branches are created? Can we say that they should be created from the release-tagged commit by whoever decides they are going to maintain that branch? Then we’ll avoid the clutter of a million release branches no one cares about.

**Comment by brandones** [comment]
This RFC is approaching one month old. I feel like it's awfully close. @openmrs/microfrontends-squad can we get some more reviews/approvals on this?

File: text/0021-versioning-releases.md
Comment (nkimaina) [reply text/0021-versioning-releases.md]: Sorry didn't see the comment on time. Do you mean something like 'An implementor or distribution that requires a certain maintenance branch, either for a minor or major release is free to create it, by branching from the release-tagged commit' ?

**Comment by nkimaina** [comment]
@ojwanganto Have you seen this? Will the proposal work for KenyaEMR?

File: text/0021-versioning-releases.md
Comment (brandones) [reply text/0021-versioning-releases.md]: Yeah, that's perfect!

**Review by brandones — APPROVED** [review]

**Review by ibacher — APPROVED** [review]

File: text/0021-versioning-releases.md
Comment (bmamlin) [inline text/0021-versioning-releases.md]: I'd suggest avoiding first person like this to simplify adoption of the RFC – i.e., write an RFC from that perspective that it has already been adopted & merged.

File: text/0021-versioning-releases.md
Comment (bmamlin) [inline text/0021-versioning-releases.md:32]: It might be worth mentioning that development should always occur on the master branch with selected changes _within master_ backported to release branches as needed. Making changes directly to a release branch and forwardporting those changes to master is discouraged and should only happen – if ever – in extreme circumstances.

Comment (dkayiwa) [reply text/0021-versioning-releases.md:32]: I completely agree!

File: text/0021-versioning-releases.md
Comment (nkimaina) [reply text/0021-versioning-releases.md]: Agreed.

File: text/0021-versioning-releases.md
Comment (nkimaina) [reply text/0021-versioning-releases.md:32]: Agreed

**Review by bmamlin — APPROVED** [review]

**Comment by brandones** [comment]
:tada:

**Pull request merged**
